Lazy and ThreadSafeLazy are implemented

viktoriia-fomina opened the pull request.

viktoriia-fomina committed:
LazyFactory, LazyThreadSafe are done

viktoriia-fomina committed:
Added method Get() to value

viktoriia-fomina committed:
I've made more comments

viktoriia-fomina committed:
A test added :)

viktoriia-fomina committed:
Added test class for multithreading tests

viktoriia-fomina committed:
added readonly to classes where i needed it. added non-threadSafe tests

viktoriia-fomina committed:
Changes

viktoriia-fomina committed:
added thread safe tests

yurii-litvinov commented on Lazy/Lazy/ILazy.cs:
```diff
@@ -0,0 +1,13 @@
+﻿
+
```
Лишние пустые строчки :)

yurii-litvinov commented on Lazy/Lazy/ILazy.cs:
```diff
@@ -0,0 +1,13 @@
+    /// </summary>
+    public interface ILazy<T>
+    {
+        /// <returns>value of the object</returns>
```
Лучше в summary написать, что Returns value of the object. Иначе summary будет пустым, что некрасиво смотрится в сгенерированной документации (если её кто-то когда-то сгенерит).

yurii-litvinov replied:
А вот тут не поправили :)

yurii-litvinov commented on Lazy/Lazy/ILazy.cs:
```diff
@@ -0,0 +1,13 @@
+    /// <summary>
+    /// ILazy interface. Implements late initialization.
+    /// </summary>
+    public interface ILazy<T>
```
По-хорошему, ILazy имеет только геттер, поэтому мог бы быть ковариантным генериком (то есть `ILazy<out T>`).

yurii-litvinov commented on Lazy/Lazy/Lazy.cs:
```diff
@@ -0,0 +1,39 @@
+        /// <summary>
+        /// Creates an object of the class Lazy.
+        /// </summary>
+        /// <param name="supplier"></param>
```
Пустые тэги не нужны

yurii-litvinov commented on Lazy/Lazy/Lazy.cs:
```diff
@@ -0,0 +1,39 @@
+        {
+            if (!isCalculated)
+            {
+                value = supplier();
```
Надо ещё сбросить supplier в null, чтобы сборщик мусора мог его собрать

yurii-litvinov commented on Lazy/Lazy/Lazy.cs:
```diff
@@ -0,0 +1,39 @@
+                value = supplier();
+            }
+
+            return value;
```
И isCalculated не поменяли :)

yurii-litvinov commented on Lazy/Lazy/LazyFactory.cs:
```diff
@@ -0,0 +1,21 @@
+        public static Lazy<T> Create(Func<T> supplier)
+        {
+            return new Lazy<T>(supplier);
+        }
```
Можно через `=>`, метод-однострочник же

yurii-litvinov commented on Lazy/Lazy/LazyThreadSafe.cs:
```diff
@@ -0,0 +1,47 @@
+
+namespace Lazy
+{
+    public class LazyThreadSafe<T> : ILazy<T>
```
Надо комментарий и к самому классу

yurii-litvinov commented on Lazy/Lazy/LazyThreadSafe.cs:
```diff
@@ -0,0 +1,47 @@
+                throw new ArgumentNullException();
+            }
+
+            Volatile.Write(ref isCalculated, false);
```
Нет нужды, конструктор одного объекта никак не вызвать из двух потоков одновременно :)

yurii-litvinov commented on Lazy/Lazy/LazyThreadSafe.cs:
```diff
@@ -0,0 +1,47 @@
+            {
+                lock (locker)
+                {
+                    if (isCalculated)
```
Тогда это чтение тоже должно быть volatile, потому что как раз в это время другой поток мог сделать запись в 35-й строчке и выйти из lock, а нас только впустили и мы видим значение в своём кеше, которое ещё false

yurii-litvinov commented on Lazy/Lazy/LazyThreadSafe.cs:
```diff
@@ -0,0 +1,47 @@
+                        return value;
+                    }
+
+                    value = supplier();
```
Тоже надо забыть про supplier

yurii-litvinov commented on Lazy/Lazy/LazyThreadSafe.cs:
```diff
@@ -0,0 +1,47 @@
+                    }
+
+                    value = supplier();
+                    isCalculated = true;
```
А тут Volatile.Write. Вроде как вход и выход из lock автоматически делают синхронизацию кешей, но Volatile.Write лучше явно написать.

yurii-litvinov commented on Lazy/LazyTests/LazyThreadSafeTests.cs:
```diff
@@ -0,0 +1,80 @@
+            }
+
+            Assert.AreEqual(1, counter);
+        }
```
Это надо и для однопоточного варианта сделать, у Вас там как раз с этим баг :)

viktoriia-fomina committed:
Added more comments and a test. Fixed some bugs

Fixed bugs:
- I added Volatile where were needed
- Since now I renew isCalculated in Lazy
- I assigned null to the supplier in Lazy and LazyThreadSafe (in order to let it to be taken by garbage collector)

yurii-litvinov reviewed (approved):
Да, теперь всё ок. Зачтена
